sycl/test-e2e/Matrix/joint_matrix_hip_apply.hpp

```diff
@@ -0,0 +1,100 @@
+
+#include <sycl/sycl.hpp>
+
+#include <cstddef>
+#include <cstdint>
+
+using namespace sycl;
+using namespace sycl::ext::oneapi::experimental::matrix;
+using sycl::ext::oneapi::bfloat16;
+
+template <typename InType, typename OutType, size_t M, size_t N, size_t K>
+void hip_matrix_apply() {
+  InType A[M * K];
+  InType B[K * N];
+  OutType C[M * N];
+  OutType D[M * N];
+  OutType E[M * N];
+
+  for (auto i = 0; i < M * K; ++i) {
+    A[i] = 1;
+  }
+
+  for (auto i = 0; i < K * N; ++i) {
+    B[i] = 2;
+  }
+
+  for (auto i = 0; i < M * N; ++i) {
+    D[i] = 0;
+    C[i] = 3;
+    E[i] = 3;
+  }
+
+  try {
+    auto defaultQueue = sycl::queue{};
+
+    auto bufA = sycl::buffer{A, sycl::range{M * K}};
+    auto bufB = sycl::buffer{B, sycl::range{K * N}};
+    auto bufC = sycl::buffer{C, sycl::range{M * N}};
+    auto bufD = sycl::buffer{D, sycl::range{M * N}};
+
+    defaultQueue
+        .submit([&](sycl::handler &cgh) {
+          sycl::accessor accA{bufA, cgh, sycl::read_write};
+          sycl::accessor accB{bufB, cgh, sycl::read_write};
+          sycl::accessor accC{bufC, cgh, sycl::read_only};
+          sycl::accessor accD{bufD, cgh, sycl::write_only};
+
+          cgh.parallel_for(
+              sycl::nd_range<2>{{4, 16}, {4, 16}}, [=](sycl::nd_item<2> idx) {
+                auto sg = idx.get_sub_group();
+                joint_matrix<sub_group, OutType, use::accumulator, M, N>
+                    sub_c{};
+                joint_matrix<sub_group, InType, use::b, K, N, layout::row_major>
+                    sub_b{};
+                joint_matrix<sub_group, InType, use::a, M, K, layout::col_major>
+                    sub_a{};
+
+                joint_matrix_load(
+                    sg, sub_a,
+                    accA.template get_multi_ptr<access::decorated::yes>(), K);
+                joint_matrix_load(
+                    sg, sub_b,
+                    accB.template get_multi_ptr<access::decorated::yes>(), N);
+                joint_matrix_load(
+                    sg, sub_c,
+                    accC.template get_multi_ptr<access::decorated::yes>(), N,
+                    layout::row_major);
+
+                joint_matrix_apply(sg, sub_a, [=](InType v) { return v * 2; });
```

Comment:
v should be passed by reference. Otherwise, sub_a will not really changed.
Please make sure you run and verify these tests

Reply:
The tests ran and verified. 
`joint_matrix_apply` lambda in HIP takes the input by value and returns the result, which is assigned to the element.
The reason for doing so, was the use of `non-array` structure to hold the underlying `joint_matrix` data.
Most of the matrix core instructions accept data declared as say `__attribute__((__vector_size__(SIZE * sizeof(TYPE))))` TYPE`
Initially, I tried to use the array to store underlying `joint_matrix` data and re-interpret cast it to the type above for matrix core instructions call. That did not work as expected. 
 
It seems I need to modify hip joint_matrix_apply to be consistent with the other backends.

Reply:
> v should be passed by reference. Otherwise, sub_a will not really changed. Please make sure you run and verify these tests

It's done in https://github.com/intel/llvm/pull/11485/commits/8bba0fb1eab9669d99fcdfecda876ba6e40cefdd

Reply:
I think the tests have not been modified. can you please check again?

Reply:
That's done @dkhaldi